system/jlib/jmetrics.hpp

```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
 };
 
 
+template <typename T>
+struct BucketInfo
+{
+    std::string label;
+    T limit;
+};
+
+
+
+template <typename T>
+struct Bucket
+{
+    Bucket(const std::string &_label, T _limit) :
+        label{_label}, limit{_limit}, count{0} {}
+    std::string label;
+    T limit;
+    RelaxedAtomic<__uint64> count;
+};
+
+
+struct BucketValue
+{
+    BucketValue(const std::string &_label, __uint64 _count) : label(_label), count(_count) {}
+    std::string label;
+    __uint64 count = 0;
+};
+
+
+template <typename T>
+struct HistogramMetricValues
+{
+    std::vector<BucketValue> bucketValues;
+    __uint64 count;
+    T sum;
+};
+
+template<typename T>
+class jlib_decl HistogramMetric : public MetricBase
+{
+public:
+    HistogramMetric(const char *name, const char *desc, StatisticMeasure _units, const std::vector<BucketInfo<T>> &bucketDefs, const MetricMetaData &_metaData = MetricMetaData()) :
+        MetricBase(name, desc, MetricType::METRICS_HISTOGRAM, _units, _metaData)
+        {
+            for (auto &b: bucketDefs)
+            {
+                buckets.emplace_back(Bucket<T>{b.label, b.limit});
+            }
+        }
+
+    virtual __uint64 queryValue() const override
+    {
+        return count;
+    }
+
+    void recordMeasurement(T measurement)
+    {
+        count.fetch_add(1);           // total cumulative count of measurements
+        sum.fetch_add(measurement);   // sum of all measurements
+        getBucket(measurement).count.fetch_add(1);   // count by buckets
+    }
+
+    HistogramMetricValues<T> queryValues()
+    {
+        HistogramMetricValues<T> values;
+        values.count = count;
+        values.sum = sum;
+        for (auto const &bucket: buckets)
+        {
+            values.bucketValues.emplace_back(BucketValue{bucket.label, bucket.count});
```

Comment:
As far as I can see the Prometheus standard is for the counts to accumulate.  i.e. if the counts are {1 ,2, 2, 1} they would be reported as { 1, 3, 5, 6 }.  Is this the responsibility of the sink or the histogram?  I don't think it matters either way, but it needs to be clearly documented.

Reply:
I opted for each bucket maintained by the metric to be the count of measurements for that bucket only. So, yes it's the responsibility of the sink to convert to the representation required. Additionally, the accumulation only needs to be done at report time reducing the overhead of recording a measurement. As far as documenting, I added a simple explanation in the metrics design doc as part of this PR.